### PR TITLE
Briefcase-138

### PR DESCRIPTION
**Pull request by batkinson**
Improved fetch performance over high-latency connections and related changes.

### CODE REVIEW

**Comment by lognaturel** [comment]
@batkinson Have you tried this with a real Aggregate server? On AppEngine?

I'm unfortunately not very well versed in this part of the system yet and have not yet understood Mitch's concern well enough to be able to gauge whether this approach avoids it. :/ Will do my best to get up to speed and provide useful feedback! I think it would be good to get several pairs of eyes on this one.

**Comment by batkinson** [comment]
Hi @lognaturel. This was tested on a production deployment of Aggregate running Tomcat on AWS. I didn't have a working AppEngine deployment to use. Mitch's concern was that low-bandwidth would have to be shared amongst multiple downloads if you go concurrent. That would actually *increase* the transmission times for those files so it could make those downloads fail where they otherwise wouldn't.

My observation here is based on running briefcase on an *actual* deployment. The briefcase download protocol is not well-designed for what it attempts to do. The http request-response style of communication for single submission file download means that latency dominates, even with high-bandwidth (like a satellite connection).

**Comment by batkinson** [comment]
Oh, one thing I forgot. I fully anticipated that the concurrency level could be configurable, which should give users the ability to configure for their environment if needed. However, my guess would be that users would probably go the *other way* and want more concurrency. I balanced this to minimize lock contention on the shared object monitors.

**Comment by lognaturel** [comment]
That's also roughly how I was understanding what Mitch said. And this is specifically a problem in the context of App Engine because of the 60-second transmission cap. However, it's not quite clear to me how likely that 60 sec limit is to be surpassed. With text-only form submissions, I'm guessing it's not common even with very low bandwidth. It's when media comes into play that things get problematic and the limit could be reached constantly. I'm not totally clear on whether that limit is inevitable, though -- [this Google reference](https://cloud.google.com/appengine/articles/deadlineexceedederrors) suggests that the server does have a chance to catch the exception though perhaps under those conditions there's not enough time. I don't know whether Aggregate tries this or not.

I think you're right that having the level of concurrency be configurable is the ideal. Mitch suggested "any concurrent-request logic needs to adjust its level of concurrency to account for the available bandwidth." I think the size of the elements being pulled would also need to be taken into account. I believe that one could get dramatically different results on the same connection and Aggregate instance depending on whether the submissions being pulled have attached media or not, for example. Does that sound correct? Do you have any media in your submissions, @batkinson?

How about adding a setting for concurrent connections that is off by default for now? I'm imagining a top-level checkbox that is something like "Try multiple connections at once" and a sub-setting that is a dropdown of plausibly reasonable choices. Then we could generate some really good help text on how to anticipate what the right value will be for each form that needs to be pulled. As more experiments happen under different conditions, the default could be changed.

If that general approach appeals to you @batkinson, same thing as with the DB change, someone else can do that piece of it on a branch that builds off this one since I know the setting is of less value to you.

**Comment by batkinson** [comment]
> With text-only form submissions, I'm guessing it's not common even with very low bandwidth. It's when media comes into play that things get problematic and the limit could be reached constantly.
Yes, I suspect the parallel downloads would only really influence things greatly when downloading multiple _large_ media attachments. One way that can could be handled is scheduling media attachments on a single thread (and http connection). However, for small media files, you're likely still better off unless you're saturating your network connection. From my testing, I couldn't get submission downloads even close to saturating my network link.

**Comment by batkinson** [comment]
I followed the advice from @lognaturel and added a setting to enable parallel downloads that is disabled by default. I also rebased this to eliminate a merge conflict after #140 was merged into master. Also, I fixed checkstyle warnings caused by IntelliJ collapsing imports from the same package into a wildcard import.

**Comment by mitchellsundt** [comment]
Things to keep in mind:

1.  Push of data to the server should be serialized.  Within ODK Aggregate, there is a **global mutex** ("TaskLock" across all server instances, mediated by the datastore layer) in the server when inserting submissions. Having multiple push requests occurring simultaneously will cause them to block on the mutex, chewing up their 60-second request limit, as they get processed in single file no matter how many server instances are spun up. 

2. That 60-second request limit can be very commonly exceeded over low-bandwidth connections, and even text-only submissions can be impacted on satellite connections. That is why ODK Collect splits submissions into multiple 10MB submission requests. The timer starts upon receipt of the first byte, so a slow connection can account for a sizeable portion of those 60 seconds. Ditto for sending a response. The processing on the server is generally negligible in relation to the transmission times.

3. Pull of data can be overlapped. However, for submissions with many media attachments, keep in mind that ODK Aggregate holds all media attachments in memory, then copies them into the HttpEntity, so this can consume a fair amount of memory (everything is held in memory at least twice). *The same memory consumption applies on pushes / submission of forms.*  I.e., you probably want to manage how many media attachment download requests you have in play at the same time.

4. Spinning up of copies of the frontend will incur faster quota usage on AppEngine. For that reason, the Aggregate configuration [here](https://github.com/opendatakit/aggregate/blob/master/eclipse-default/WebContent/WEB-INF/appengine-web.xml) specifies a 14-second queuing time threshold before a new instance is spun up.  Only if at least one request is queued for longer than 14 seconds will a new instance be spun up, and then that new instance will take about 30 seconds to become live. Leaving a 15-second processing interval. This is why ODK Collect tried twice before failing a submit.

The two worst of these limitations, the global mutex and the in-memory copies/full-packet-assembly, are a result of implementing on top of AppEngine and its Datastore. 

A server that used database transactions and that used streaming servlet 3.0 functionality would have less trouble with concurrent requests.

File: src/org/opendatakit/briefcase/util/WebUtils.java [hunk 1]
```diff
@@ -307,39 +315,37 @@ public static final HttpPost createOpenRosaHttpPost(URI uri) {
...
+
+		HttpClientBuilder clientBuilder = HttpClientBuilder.create()
+				.setConnectionManager(connectionManager)
+				.setConnectionManagerShared(true)
+				.setDefaultSocketConfig(socketConfig)
+				.setDefaultRequestConfig(requestConfig);
+
+		HttpHost proxy = BriefcasePreferences.getBriefCaseProxyConnection();
+		if (proxy != null) {
+			clientBuilder.setRoutePlanner(new DefaultProxyRoutePlanner(proxy)).build();
```
Comment (lognaturel) [inline src/org/opendatakit/briefcase/util/WebUtils.java]: I think this call on `build` is redundant. Not sure whether it matters.

File: src/org/opendatakit/briefcase/util/ServerFetcher.java [hunk 2]
```diff
@@ -238,88 +250,175 @@ public boolean downloadFormAndSubmissionFiles(List<FormStatus> formsToTransfer)
...
+        for (String uri : chunk.uriList) {
+          if (isCancelled()) {
+            fs.setStatusString("aborting requesting submissions...", true);
+            EventBus.publish(new FormStatusEvent(fs));
+            return false;
+          }
+          submissionCompleter.submit(new SubmissionDownload(formInstancesDir, formDatabase, lfd, fs, uri));
+        }
+
+        for (int i = 0; i < chunk.uriList.size(); i++) {
```
Comment (lognaturel) [inline src/org/opendatakit/briefcase/util/ServerFetcher.java]: I think this can be replaced with a for-each loop. It confused me since there's the same loop above so I was trying to figure out a reason for it being different. I'm apparently easily confused!

File: src/org/opendatakit/briefcase/ui/MessageStrings.java [hunk 3]
```diff
@@ -80,4 +80,6 @@
   public static final String PROXY_SET_ADVICE =
       "If you are behind a proxy, try setting up your proxy details through 'Settings' tab.";
 
+  public static final String PARALLEL_DOWNLOADS = "Use Parallel Downloads";
```
Comment (lognaturel) [inline src/org/opendatakit/briefcase/ui/MessageStrings.java]: I think it would be good to indicate that this is for pulls only.

File: src/org/opendatakit/briefcase/util/ServerFetcher.java (hunk 2 above)
Comment (lognaturel) [inline src/org/opendatakit/briefcase/util/ServerFetcher.java]: Or perhaps you did it this way to make it clear that no elements are actually used and that it's purely a counting loop? That actually makes sense.

**Comment by lognaturel** [comment]
Really appreciate your insights, @mitchellsundt, those are so helpful. From what you say, I think this level of parallelism will mostly be helpful for those who don't have media attachments and/or don't run on AppEngine. The point about in-memory copies is particularly interesting and it seems that might be a good next step for refining this implementation. 

When we release this, we really need to be thoughtful about how we communicate to users when parallelism might help them and when it could hurt them.

File: src/org/opendatakit/briefcase/util/ServerFetcher.java (hunk 2 above)
Comment (batkinson) [inline src/org/opendatakit/briefcase/util/ServerFetcher.java]: Yes, I didn't use the for-each here because the second loop doesn't actually use the items in the list. It only needs to match up with the first loop, since each submitted job in the first loop should have a matching call to `take()` in the second. I can add a comment to clarify and prevent the reader from having to dive directly into the loop semantics.

File: src/org/opendatakit/briefcase/util/WebUtils.java (hunk 1 above)
Comment (batkinson) [inline src/org/opendatakit/briefcase/util/WebUtils.java]: Yes, you're right. At the very least, it's confusing. I just missed this when converting the original code. I'll remove it.

File: src/org/opendatakit/briefcase/ui/MessageStrings.java (hunk 3 above)
Comment (batkinson) [inline src/org/opendatakit/briefcase/ui/MessageStrings.java]: Ok, I thought download implied that, but it's the only place where the word download exists in the UI. I will reword this to make it more clear.

**Comment by yanokwa** [comment]
Nice work, @batkinson! I've confirmed that this is a lot faster on a small form with no attachments on App Engine and a local Tomcat server. I've also confirmed that it works via an HTTP proxy.

Marking parallel downloads as an experimental feature is a good way to communicate to users that the feature might not work for them, so great job there too!
